scipy/spatial/ckdtree/src/build.cxx

```diff
@@ -142,62 +127,259 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+        std::copy_n(maxes, maxes_.size(), maxes_.begin());
+
+        future_ = std::async(std::launch::async, [&]() {
+                buflist_.emplace_back();
+                buflist_.back().reserve((end_idx-start_idx)/self->leafsize/workers);
+                return build(
+                    self, start_idx, end_idx, maxes_.data(), mins_.data(), _median, _compact,
+                    &buflist_, subworkers);
+            });
+
+        return subworkers;
+    }
 
+    operator bool() const { return future_.valid(); }
+    void wait() { future_.wait(); }
+    std::list<std::vector<ckdtreenode>>& buflist() { return buflist_; }
+};
+
+static ckdtree_intp_t
+build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
+      double *maxes, double *mins,
+      const int _median, const int _compact,
+      std::list<std::vector<ckdtreenode>>* buflist, int workers)
+{
+    std::vector<ckdtreenode>& buf = buflist->front();
+
+    if (workers == 1)
+        return build(self, start_idx, end_idx, maxes, mins, _median, _compact, buf);
+
+    /* put a new node into the node stack */
+    const ckdtree_intp_t node_index = buf.size();
+    ckdtree_intp_t _less;
+    buf.emplace_back();
+    ckdtreenode *node = &buf.back();
+    init_ckdtreenode(*node, start_idx, end_idx);
+
+    if (end_idx-start_idx <= self->leafsize) {
+        /* below brute force limit, return leafnode */
+        node->split_dim = -1;
+        return node_index;
+    }
+    else {
+
+        const ckdtree_intp_t p = split_and_partition(self, start_idx, end_idx, maxes, mins,
+            _median, _compact, node);
+        if(p < 0)
+            return node_index; // split_and_partition decided to have a leaf node
+
+        const double split = node->split;
+        const ckdtree_intp_t d = node->split_dim;
+
+        // try to put this on its own cache line to avoid false sharing with the child thread
+        alignas(64) build_worker worker;
+
+        if (CKDTREE_LIKELY(_compact)) {
+            workers -= worker.run(self, p, end_idx, maxes, mins, _median, _compact, workers);
+            _less = build(self, start_idx, p, maxes, mins, _median, _compact, buflist, workers);
+        }
+        else
+        {
+            std::vector<double> mids(self->m);
+
+            std::copy_n(mins, mids.size(), mids.begin());
+            mids[d] = split;
+            workers -= worker.run(self, p, end_idx, maxes, mids.data(), _median, _compact, workers);
+
+            std::copy_n(maxes, mids.size(), mids.begin());
+            mids[d] = split;
+            _less = build(self, start_idx, p, mids.data(), mins, _median, _compact, buflist, workers);
+        }
+
+        node = &buf[node_index];
+        /* fill in entries */
+        node->_less = _less;
+
+        ckdtree_intp_t _greater = 0;
+        for(const auto& b : *buflist)
+            _greater += b.size();
+        worker.wait();
+        buflist->splice(buflist->end(), worker.buflist());
+        node->_greater = _greater;
+
```

Comment:
```node->split_dim``` and ```node->split``` are not filled in

Reply:
The are set by `split_and_partition`, here:
https://github.com/scipy/scipy/blob/017f9a329196a0a4026805607c5e2230494e8121/scipy/spatial/ckdtree/src/build.cxx#L146-L147